### PR TITLE
macos: Add updater support

### PR DESCRIPTION
**Pull request by marysaka**
This is a very basic updater but should be enough for now.

As we cannot replace ourself while running (otherwise we could get killed because our codesigning get invalidated), the update flow is a bit different. We use a bash script to replace the application and relaunch it.

I also did a bit of refactoring around to make InstallUpdate more readable.

### CODE REVIEW

**Comment by marysaka** [comment]
Rebased

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 1]
```diff
@@ -381,6 +398,15 @@ private static void DoUpdateWithMultipleThreads(TaskDialog taskDialog, string do
 
                         File.WriteAllBytes(updateFile, mergedFileBytes);
 
+                        // On macOS ensure we remove the quarantine bit so we don't have Gatekeeper being annoying.
+                        if (OperatingSystem.IsMacOS())
+                        {
+                            using (Process xattrProcess = Process.Start("xattr", new List<string> { "-d", "com.apple.quarantine", updateFile}))
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:404]:
```suggestion
                            using (Process xattrProcess = Process.Start("xattr", new List<string> { "-d", "com.apple.quarantine", updateFile }))
```

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 2]
```diff
@@ -286,22 +287,38 @@ private static async void UpdateRyujinx(Window parent, string downloadUrl)
...
 
-                    if (!Path.Exists(ryuExe))
+                    if (!Path.Exists(executablePath))
                     {
-                        ryuExe = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, OperatingSystem.IsWindows() ? "Ryujinx.exe" : "Ryujinx");
+                        executablePath = Path.Combine(executableDirectory, OperatingSystem.IsWindows() ? "Ryujinx.exe" : "Ryujinx");
                     }
 
-                    Process.Start(ryuExe, CommandLineState.Arguments);
+                    // On macOS we perform the update at relaunch
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:311]:
```suggestion
                    // On macOS, we perform the update at relaunch.
```

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 3]
```diff
@@ -286,22 +287,38 @@ private static async void UpdateRyujinx(Window parent, string downloadUrl)
...
+                    if (OperatingSystem.IsMacOS())
+                    {
+                        string baseBundlePath = Path.GetFullPath(Path.Combine(executableDirectory, "..", ".."));
+                        string updaterScriptPath = Path.Combine(UpdateDir, "Ryujinx.app", "Contents", "Resources", "updater.sh");
 
+                        executablePath = "/bin/bash";
+                        arguments.InsertRange(0, new List<string> { updaterScriptPath, baseBundlePath, UpdateDir});
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:318]:
```suggestion
                        arguments.InsertRange(0, new List<string> { updaterScriptPath, baseBundlePath, UpdateDir });
```

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 4]
```diff
@@ -470,87 +496,96 @@ private static void DoUpdateWithSingleThread(TaskDialog taskDialog, string downl
...
-                    }
+                File.SetLastWriteTime(outPath, DateTime.SpecifyKind(zipEntry.DateTime, DateTimeKind.Utc));
+
+                Dispatcher.UIThread.Post(() =>
+                {
+                    taskDialog.SetProgressBarState(GetPercentage(count, zipFile.Count), TaskDialogProgressState.Normal);
                 });
             }
+        }
+
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:566]: Double empty line.

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 5]
```diff
@@ -560,38 +595,42 @@ private static async void InstallUpdate(TaskDialog taskDialog, string updateFile
             taskDialog.SubHeader = LocaleManager.Instance[LocaleKeys.UpdaterRenaming];
             taskDialog.SetProgressBarState(0, TaskDialogProgressState.Normal);
 
-            // Replace old files
-            await Task.Run(() =>
+            // NOTE: On macOS replacement is delayed to the restart phase.
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:598]:
```suggestion
            // NOTE: On macOS, replacement is delayed to the restart phase.
```

File: distribution/macos/updater.sh [hunk 6]
```diff
@@ -0,0 +1,15 @@
+#!/bin/bash
+
+set -e
+
+INSTALL_DIRECTORY=$1
+NEW_APP_DIRECTORY=$2
+APP_ARGUMENTS="${@:3}"
+
+# Wait 1s to give Ryujinx some time to exit.
+sleep 1
```
Comment (gdkchan) [inline distribution/macos/updater.sh:10]: This seems fragile. What if it takes more than one second to exit?

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 7]
```diff
@@ -470,87 +496,96 @@ private static void DoUpdateWithSingleThread(TaskDialog taskDialog, string downl
...
-                    {
-                        while ((tarEntry = tarStream.GetNextEntry()) is not null)
-                        {
-                            if (tarEntry.IsDirectory) continue;
+            TarEntry tarEntry;
 
-                            string outPath = Path.Combine(UpdateDir, tarEntry.Name);
+            while ((tarEntry = tarStream.GetNextEntry()) is not null)
+            {
+                if (tarEntry.IsDirectory) continue;
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:511]:
```suggestion
                if (tarEntry.IsDirectory)
                {
                    continue;
                }
```

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 8]
```diff
@@ -560,38 +595,42 @@ private static async void InstallUpdate(TaskDialog taskDialog, string updateFile
...
+                            Dispatcher.UIThread.Post(() =>
+                            {
+                                taskDialog.SetProgressBarState(GetPercentage(count, allFiles.Count), TaskDialogProgressState.Normal);
+                            });
+                        }
+                        catch
                         {
-                            taskDialog.SetProgressBarState(GetPercentage(count, allFiles.Count), TaskDialogProgressState.Normal);
-                        });
+                            Logger.Warning?.Print(LogClass.Application, LocaleManager.Instance.UpdateAndGetDynamicValue(LocaleKeys.UpdaterRenameFailed, file));
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:619]: Interesting, I didn't know we had localization for log messages.

File: distribution/macos/updater.sh (hunk 6 above)
Comment (marysaka) [reply distribution/macos/updater.sh:10]: Well then it will get killed by codesigning here

File: distribution/macos/updater.sh [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+# Wait for Ryujinx to exit
+# NOTE: in case no fds are open, lsof could be returning with a process still living.
+# We wait 1s and assume the process stopped after that
+lsof -p $APP_PID +r 1 &>/dev/null
+sleep 1
+
+# Now replace and reopen.
+rm -rf $INSTALL_DIRECTORY
+mv $NEW_APP_DIRECTORY $INSTALL_DIRECTORY
+open -a $INSTALL_DIRECTORY --args $APP_ARGUMENTS
```
Comment (TSRBerry) [inline distribution/macos/updater.sh:39]:
```suggestion
rm -rf "$INSTALL_DIRECTORY"
mv "$NEW_APP_DIRECTORY" "$INSTALL_DIRECTORY"
open -a "$INSTALL_DIRECTORY" --args "$APP_ARGUMENTS"
```
Just to be safe

**Review by TSRBerry — COMMENTED** [review]
Updater code lgtm! I only have one small suggestion for the updater script.

**Review by TSRBerry — APPROVED** [review]
Lgtm!

File: Ryujinx.Ava/Modules/Updater/Updater.cs [hunk 10]
```diff
@@ -286,22 +287,41 @@ private static async void UpdateRyujinx(Window parent, string downloadUrl)
...
 
-                    Process.Start(ryuExe, CommandLineState.Arguments);
+                    // On macOS we perform the update at relaunch.
+                    if (OperatingSystem.IsMacOS())
+                    {
+                        string baseBundlePath = Path.GetFullPath(Path.Combine(executableDirectory, "..", ".."));
+                        string newBundlePath = Path.Combine(UpdateDir, "Ryujinx.app");
+                        string updaterScriptPath = Path.Combine(newBundlePath, "Contents", "Resources", "updater.sh");
+                        string currentPid = Process.GetCurrentProcess().Id.ToString();
+
```
Comment (gdkchan) [inline Ryujinx.Ava/Modules/Updater/Updater.cs:318]: Double empty line.

**Review by AcK77 — APPROVED** [review]

**Pull request merged**